user activation link doesn't enforce frontend base URL

lcdservices opened the pull request:
#7737 - enforce frontend construction for user activation URL

https://github.com/joomla/joomla-cms/issues/7737

---

when the register method constructs the user activation URL, it doesn't explicitly enforce a frontend URL. this means that 3rd party extensions wanting to trigger registration (and thus activation emails) from a backend tool cannot make use of the core register method.

since the activation URL is only valid from the frontend, we should always enforce the proper construction.

there are two locations to be modified:
https://github.com/joomla/joomla-cms/blob/staging/components/com_users/models/registration.php#L410
https://github.com/joomla/joomla-cms/blob/staging/components/com_users/models/registration.php#L447

In both cases, we can just include the base url inside JRoute::_() to achieve the enforcement:

$data['activate'] = JRoute::_($base . '/index.php?option=com_users&task=registration.activate&token=' . $data['activation'], false);

lcdservices committed:
Update registration.php

#7737 - enforce frontend construction for user activation URL

zero-24 commented:
@lcdservices I have just tested this but it don't work for me.

### How to test
- enable frontend registratration
- choose "self" registration
- go to the frontend and register an account
- click the activation link. --> Works
- got to the backend and set "admin registration"
- go to the frontend and register an account
- click the activation link. --> Works
- apply this patch

- choose "self" registration
- go to the frontend and register an account
- click the activation link. --> don't work here
- got to the backend and set "admin registration"
- go to the frontend and register an account
- click the activation link. --> don't work here.

### expected result

`http://www.example.org/dev/e3/index.php?option=com_users&task=registration.activate&token=ab7bc2b688beb2477c103ce4ee5b318b`

### actual result

`http://www.example.org/index.php?option=com_users&task=registration.activate&token=ab7bc2b688beb2477c103ce4ee5b318b`

infograf768 commented:
I wonder if we need $base here.
Maybe we would just need `JRoute::_(JUri::root() . ''index.php?option=com_users&task=registration.activate&token=' . $data['activation'], false);

joomdonation commented:
This code is wrong. And I think the solution is quite simple. We just need to remove administrator/ from $data['activate'] when this method is called from administrator area with this code

```php
if (JFactory::getApplication()->isAdmin())
{
	$adminPos         = strrpos($data['activate'], 'administrator/');
	$data['activate'] = substr_replace($data['activate'], '', $adminPos, 14);
}
```

I made a small PR to your branch to make it works like that https://github.com/lcdservices/joomla-cms/pull/1. Please try to test it and if it works, please merge it so that other testers can test it

lcdservices committed:
adjust how the root URL is created

use JUri::root() as the base URL rather than constructing via JUri::getInstance()->toString()

lcdservices commented:
I think the suggestion from @infograf768 is much cleaner and more direct than doing a string replacement after the URL has been constructed. @zero-24 -- can you test the revised PR? I don't have a dev site set up where Joomla is installed in a subdirectory (which is the use case that created the issues you noticed).

joomdonation commented:
@lcdservices While the suggestion from @infograf768 works, it always return none-sef URL. Also, I haven't seen anywhere in our core code which pass full URL into JRoute::_ method, so I don't think that the code is good enough.

zero-24 commented:
hmm generaly it works but it looks like we need to fix is also here:
https://github.com/lcdservices/joomla-cms/blob/patch-1/components/com_users/models/registration.php#L85

We also get with this patch applyed:

![error_registratration](https://cloud.githubusercontent.com/assets/2596554/9493087/7d45cdda-4bfd-11e5-9ac6-015fc0410b5f.JPG)

But i have no clue why?

zero-24 commented:
hmm i guess the seccond on was an error on my side. I can not reproduce this again?!

Ok. So only point on need to be fixed see:
https://github.com/lcdservices/joomla-cms/blob/patch-1/components/com_users/models/registration.php#L85

and the comment by @joomdonation

joomdonation commented:
@zero-24 I think we don't have to worry at the code at activate method (which you mentioned). The reason is because usually, third party developer only uses register method to create a Joomla user in Joomla core (I use it in my extension too), but for the activate method, I think it is just handled by Joomla core (when users click on the activation link).

lcdservices committed:
Update registration.php

lcdservices committed:
Update registration.php

lcdservices committed:
Update registration.php

lcdservices commented:
ok...
@zero-24 I fixed it in the additional location you identified
@joomdonation - I really don't like the string replace method -- it's too hackish. but I agree it seems the only way to do this while preserving SEF urls. So I restructured it to handle it in this way.

zero-24 committed:
let travis be a happy guy

fixes cs and hopfully make travis happy

joomdonation commented:
@lcdservices Could you merge @zero-24 PR so that travis happy and we can get this PR merged?

I use this function in my extension, too, and would love to have this merged in next release of Joomla so that I don't have to change code / fix bugs for client

lcdservices committed:
Merge pull request #2 from zero-24/patch-8

let travis be a happy guy

lcdservices commented:
done. sorry, didn't notice he had done it as a PR against my patch.

zero-24 commented:
RTC based on testing :smile: <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/7765">issues.joomla.org/joomla-cms/7765</a>.</sub>

infograf768 commented:
Any reason not to use a simpler:
`$data['activate'] = str_replace( 'administrator/', '', $data['activate'])`
instead of looking for its position?

joomdonation commented:
Might have bug in case you you have a site called administrator

for example, http://localhost/administrator/administrator

Of course it will only happens in that very special case but we should avoid that.
